### ISSUE TITLE
Rolemapping only accept principalId where with like

### ISSUE THREAD

**Issue by felipetoffolo1** [strongloop/loopback#3121]
### Bug/Feature request

After i upgraded Loopback to version 3, This filter o rolemapping returns nothing
`{"where":{"principalId":"5880038427758463a1ae5a3b"}}`
i changed my filter to this
`{"where":{"principalId":{ "like": "5880038427758463a1ae5a3b"}}}`
and it worked. But is it a bug? Or it changed onn the new version? (i did not find anything about it in migration docs)

### Expected result
Expected
filter` {"where":{"principalId":"5880038427758463a1ae5a3b"}`} to return results

Actual
it returns nothing

### Additional information
linux x64 7.4.0

├── loopback@3.2.1
├── loopback-boot@2.23.0
├── loopback-component-explorer@4.0.0
├── loopback-component-push@1.7.0
├── loopback-component-storage@3.0.0
├── loopback-connector-mongodb@3.0.0

**Comment by alFReD-NSH** [comment]
Try converting `"5880038427758463a1ae5a3b"` to object id.

**Comment by totalard** [comment]
How i do that?

**Comment by luncht1me** [comment]
Converting to ObjectID doesn't work. I'm having the same problem.

Trying to find a RoleMapping using principalId in the where query returns nothing.
If I try and find it as a regex of the Id string it works.
principalId is stored as a string in mongodb.

**Comment by luncht1me** [comment]
It's kind of annoying, I had a nice function to return the roles of a user based on their mapping
```javascript
RoleMapping.find({
	where:{
		"principalType": RoleMapping.USER,
		"principalId": userId.toString()
	}
}, function (err, roleMap) {
	console.log("[getRolesForUser]:", err, roleMap);
	if (err || (roleMap && !roleMap.length)) {
		cb("no rolemapping found", false);
	}
	else if (roleMap && roleMap.length) {
		/* Map out the roleIds only. Some reason can't filter fields above */
		var roleIds = roleMap.map(function (r) {
			return r.roleId
		});
		console.log("[getRolesForUser - roleIds]: ", roleIds);
		/* Get our roles that match the id map */
		Role.find({
			where: {
				id: {
					inq: roleIds
				}
			}

		}, function (err, roles) {
			console.log("[Role.find]:", err, roles);
			if (err || !roles.length) {
				cb("no roles found", false);
			} else {
				/* Map out only the names of the roles found */
				var roleNames = roles.map(function (r) {
					return r.name;
				});
				/* Return roleNames */
				cb(null, roleNames);
			}
		})
	}
});
```

But now this has to be changed since doing the initial `where` with `principalId` returns nothing....

The resulting function is much less verbose, as I can only realistically target a single Role, and then check thru all the mappings that our userId is included.
It's ugly. It's bad. I hate it.

```javascript
Role.findOne({
	where:{
		name: "$admin"
	}
}, function(err, role){
	if (err || !role) cb(err, "no admin role found");
	RoleMapping.find({
		where:{
			roleId: role.id
		}
	}, function(err, rm){
		console.log(err, rm);
		if(err || !rm) cb(err, "no role mapping found");
		var ids = rm.map(function(r){
			return r.principalId
		});
		console.log(ids);
		if(ids.indexOf(userId.toString()) !== -1){
			cb(null, true);
		} else {
			cb(null, false);
		}
	});
});
```

Please fix principalId where queries!

To recap;

`where:{
  principalId: "exactIdMatch"
}` returns nothing, but `where:{ principalId: /idString/ }` does return the document, but is very dangerous since an incomplete idString could easily return a different user's rolemapping.
principalId's type in the database is indeed a string.

**Comment by luncht1me** [comment]
Just a debug output:
```
  loopback:connector:mongodb all +2ms RoleMapping { where:
   { roleId: 589b917cc889c618f09ca889,
     principalId: '589b917cc889c618f09ca888' } }
  loopback:connector:mongodb MongoDB: model=RoleMapping command=find +2ms [ { roleId: 589b917cc889c618f09ca889,
    principalId: 589b917cc889c618f09ca888 },
  [Function] ]
```

In the first output, the principalId is a string, as seen by the quotes. But then the next output, principalId is no longer a string but an actual Id. I'm assuming this casting here is what's breaking the query.

**Comment by luncht1me** [comment]
After lots of debugging. I've come to a solution.

In loopback's default `common/models/role-mapping.json`, added `"strictObjectIDCoercion": true` to the model definition.
Now principalId can be queried properly.

resulting `node_modules/loopback/common/models/role-mapping.json`:
```json
{
  "name": "RoleMapping",
  "description": "Map principals to roles",
  "strictObjectIDCoercion": true,
  "properties": {
    "id": {
      "type": "string",
      "id": true,
      "generated": true
    },
    "principalType": {
      "type": "string",
      "description": "The principal type, such as user, application, or role"
    },
    "principalId": {
      "type": "string",
      "index": true
    }
  },
  "relations": {
    "role": {
      "type": "belongsTo",
      "model": "Role",
      "foreignKey": "roleId"
    }
  }
}
```

**Comment by alFReD-NSH** [comment]
Seems like same issue at #1441. You can also try my workaround which I described in [here](https://github.com/strongloop/loopback/issues/1441#issuecomment-180767939):

```
var ObjectID = RoleMapping.getDataSource().connector.getDefaultIdType();

// Because of this: https://github.com/strongloop/loopback-connector-mongodb/issues/1441
RoleMapping.defineProperty('principalId', {
  type: ObjectID,
});
```

**Comment by luncht1me** [comment]
this requires overwriting the internal RoleMapping model. We shouldn't have
to do this for default intended functionality.

The fact remains theres a discrepancy and inconsistency between
RoleMapping's definition of principalId and how it's being handled in
queries.

**Comment by luncht1me** [comment]
but thankyou for the alternate work around.

On Thu, Feb 9, 2017, 11:38 PM Colin Johnson, <lunchtime@nxtlvl.gg> wrote:

this requires overwriting the internal RoleMapping model. We shouldn't have
to do this for default intended functionality.

The fact remains theres a discrepancy and inconsistency between
RoleMapping's definition of principalId and how it's being handled in
queries.

On Thu, Feb 9, 2017, 8:11 PM Farid Nouri Neshat, <notifications@github.com>
wrote:

Seems like same issue at #1441
<https://github.com/strongloop/loopback/issues/1441>. You can also try my
workaround which I described in here
<https://github.com/strongloop/loopback/issues/1441#issuecomment-180767939>:

var ObjectID = RoleMapping.getDataSource().connector.getDefaultIdType();

// Because of this:
https://github.com/strongloop/loopback-connector-mongodb/issues/1441
RoleMapping.defineProperty('principalId', {
  type: ObjectID,
});

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/strongloop/loopback/issues/3121#issuecomment-278853028>,
or mute the thread

**Comment by Casy** [comment]
any update on this?

**Comment by Casy** [comment]
I try with @luncht1me solutions but still have this issue.

And when i try with like:
`Error: The principalId property has invalid clause {"like":"58a343ec165687d52aa62bc9"}: Expected a string or RegExp
`

**Comment by bajtos** [comment]
cc @fabien I know you are using the MongoDB connector a lot, have you encountered this issue too? Can you think of any advice on how to fix it in user-land?

**Comment by bajtos** [comment]
> Seems like same issue at #1441

Can we close this issue as a duplicate of #1441 then? /cc @alFReD-NSH

**Comment by fabien** [comment]
@bajtos here's what I have in a bootscript - it works fine:

```
var RoleMapping = app.models['RoleMapping'];
RoleMapping.settings.strictObjectIDCoercion = true;
```

**Comment by bajtos** [comment]
Hello, I submitted https://github.com/strongloop/loopback-workspace/pull/437 that will enable `strictObjectIDCoercion` for all new applications scaffolded using our `lb`/`apic`/`slc` tooling.

For existing applications, you can either add a boot script as @fabien shown above, or you can enable the setting via `server/model-config.json`:

```json
{
  "RoleMapping": {
    "dataSource": "db",
    "options": {
      "strictObjectIDCoercion": true
    },
    "public": false
  }
}
```

As part of the update, according to https://github.com/strongloop/loopback/pull/3198#issuecomment-280382924, you also need to update the data already stored in your MongoDB database to fix the type of the relevant id fields.

**Comment by bajtos** [comment]
I am closing this issue as resolved.

**Comment by pr3mar** [comment]
This error still exists.

**Comment by bajtos** [comment]
@pr3mar 
> This error still exists.

Please open a new GitHub issue and provide us with instruction how to reproduce the error you are seeing, see http://loopback.io/doc/en/contrib/Reporting-issues.html#bug-report
